### PR TITLE
Add Operator to Update Indexed Values

### PR DESCRIPTION
**Pull request by lara-hdr**

### CODE REVIEW

**Comment by lara-hdr** [comment]
I could modify the operator's name for "UpdateIndex" or "Update" or other

**Comment by gramalingam** [comment]
How does this op related to Scatter (https://github.com/onnx/onnx/blob/master/docs/Operators.md#Scatter ) ? Will Scatter be sufficient for the current purpose?

**Comment by lara-hdr** [comment]
@gramalingam, both this op and scatter would be used to update values in an input tensor.
However, Scatter has more restrictions that make it impossible to cover all cases of this operator.
Scatter follows an equation to update the values (for example, in a 2-D tensor case, data[indices[i][j]][j] = updates[i][j] with an axis of 0); this makes it hard to update some of the values without updating others.

Let's assume data is : 
 [ [ 1, **2**, **3**],
   [ 4, 5, **6**],]
and we want to update the bold values.

File: onnx/defs/tensor/defs.cc [hunk 1]
```diff
@@ -691,6 +691,117 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+ONNX_OPERATOR_SET_SCHEMA(
+    IndexPut,
+    10,
+    OpSchema()
+        .SetDoc(IndexPut_ver10_doc)
+        .Input(0, "data", "Tensor of rank r >= 1.", "T")
+        .Input(
+            1,
+            "indices",
+            "Tensor of int32/int64 indices of rank q <=r, can be the indexes of a particular element or a slice (Example 2).",
```
Comment (wschin) [inline onnx/defs/tensor/defs.cc]:
```suggestion
            "Tensor of int32/int64 indices of rank q <= r, can be the indexes of a particular element or a slice (Example 2).",
```

File: onnx/defs/tensor/defs.cc [hunk 2]
```diff
@@ -691,6 +691,117 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+      [1, 1, 1],
+      [2, 2, 2],
+      [0, 0, 0],
+      [0, 0, 0],
+      [3, 3, 3]
+  ]
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    IndexPut,
```
Comment (wschin) [inline onnx/defs/tensor/defs.cc]: I am not sure if `IndexPut` is a good name. How about `Replace`? It's closer to what it's doing in your examples.

Comment (lara-hdr) [reply onnx/defs/tensor/defs.cc]: That's a good name. I'll change the name for Replace

File: onnx/backend/test/case/node/index_put.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
+from __future__ import absolute_import
```
Comment (wschin) [inline onnx/backend/test/case/node/index_put.py]: Please also rename this file.

File: onnx/backend/test/case/node/index_put.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        updates = np.array([1, 2, 3], dtype=np.float32)
+
+        y = np.array([[0, 1, 0],
+                     [0, 0, 2],
+                     [0, 0, 0],
+                     [0, 0, 0],
+                     [0, 3, 0]], dtype=np.float32)
+
+        expect(node, inputs=[data, indices, updates], outputs=[y],
+               name='test_index_put')
```
Comment (wschin) [inline onnx/backend/test/case/node/index_put.py]: Test names are not renamed.

File: onnx/backend/test/case/node/index_put.py (hunk 3 above)
Comment (lara-hdr) [reply onnx/backend/test/case/node/index_put.py]: The file is already renamed, this code is outdated :)

File: onnx/backend/test/case/node/index_put.py (hunk 4 above)
Comment (lara-hdr) [reply onnx/backend/test/case/node/index_put.py]: The file is already renamed, this code is outdated :)

**Pull request closed without merging**